### DISCUSSION TITLE
Upload ROM and newest savegame at the same time

### DISCUSSION THREAD

**Discussion by Dakkaron** [sanni/cartreader#182] [Ideas & Suggestions]
I am using the cartreader and a chinese repo cartrige to swap through multiple ROMs that I like to play. To keep the savegames I always have to flash the ROM and then write the savegame.

I think it would be nice to have the option to automatically fetch the latest savegame and write it to the cartridge.

The flow that I imagine is as follows:

When I select a ROM with a path that starts with `/GB/ROM/`, e.g. `/GB/ROM/MY_GAME/5/MYGAME.GB` it checks if a folder with the name `/GB/SAVE/MY_GAME/` exists. If it does, it finds the sub-folder with the highest number and uploads the savegame inside.

This would only work when flashing ROMs under `/GB/ROM/`. But it would make it much quicker to swap between games and their savegames.

**Comment by sanni** [comment]
Hi,

I have seen your Pull request, did you test it with an MBC5 cart like this? https://www.jrodrigo.net/wiki/how-to-diy-a-gameboy-flash-cartridge-with-a-rom-adapter-board/
I can't seem to flash mine with your code change, could be just me ofc.

**Reply by Dakkaron** [reply]
I don't have one. I just went off the documentation and the previous code.

I only found three differences between MBC3 and MBC5 regarding flashing:

- In MBC5 bank 0 can be mapped like the other banks (but still should be mapped seperatly like in MBC3)
- For MBC5 this line is executed before flashing. I don't know what it does, but it shouldn't matter for MBC3: `writeByte_GB(0x4000, 0x0);`
- MBC5 sets the current bank numer at address 0x2000, while MBC3 does it at 0x2100.

For the first one I went with the MBC3 way, since it should be compatible with both.
For the second one, I went with MBC5, because MBC3 should ignore the write.
For the third one, I went with MBC3, because I couldn't see the difference and chose at random. I guess, the 3rd one will be what causes the problem. Maybe changing that to 0x2100 would solve the problem. Or otherwise, writing the bank number to both addresses.

The only flashcart I have for testing is a chinese repro cartridge (SD007_TSOP_29LV017D). I added support for that cartridge, but I couldn't figure out how to identify the cartridge from the software (the ID command always returns different values), so I haven't submitted that yet.

**Reply by Dakkaron** [reply]
I pushed a change that might fix flashing for MBC5. Could you please try if it works?

**Comment by sanni** [comment]
Same problem, it freezes at "Writing flash MBC3/MBC5" with no blinking of the LED.

The flashing fails at different offsets for different roms but always at the same offset for the same rom, and always within the first 512 bytes:

When I remove those two lines the flashing works:

![484841511141411](https://user-images.githubusercontent.com/19169494/124757959-691e1180-df2e-11eb-98fb-b427c9e87f23.jpg)

**Reply by Dakkaron** [reply]
Interesting... In MBC3 rom bank 0 is only accessible via addresses `0x0000-0x3FFF`, with banks 1 and higher being addressible under `0x4000-0x7FFF`.
In MBC5 it is the same, only that bank 0 is also accessible via addresses `0x4000-0x7FFF`. So when flashing MBC3, you need to write the first bank starting at address `0x0000`, while all other banks are written starting at `0x4000`. Thats what that code does.
When flashing MBC5, it should be workable the same way, but apparently it is not...

I guess flashing gets stuck in the busy check? Other than that I don't see any other place where it could get stuck.

One way to handle that would be to try with MBC3-style flashing. Then we add a timeout to the busy code and if the timeout is hit (e.g. 1000 cycles of the busy check), restart flashing MBC5-style.

What do you think?

**Reply by Dakkaron** [reply]
Have a look at the PR here: https://github.com/sanni/cartreader/pull/183
The latest commits should fix that issue.
